DOCS-2930: Add first_run flow for Docker modules (#4187)
viamrobotics/docs: docs/operate/reference/advanced-modules/_index.md

However, if you are unable to use modular resources because you need to host `viam-server` on a non-Linux system or have an issue with compilation, you may need to [implement a custom component and register it on a server configured as a remote](/operate/reference/advanced-modules/custom-components-remotes/) on your machine.

> ## Package and deploy using Docker

Reviewer: For both Custom APIs and components as remote parts, we have subpages. I think we should do the same here.

> In rare cases, you may need to package and deploy a module using Docker.

Reviewer: Be aware that we are not providing complete info here. We're showing how to pull and run a container. We will likely need to look into how to package a module with Docker. So far this really only covers the deploy aspect.

Please file a follow up ticket to cover this. A full example for reference is really what I think we need here.

Author: https://viam.atlassian.net/browse/DOCS-3863

However, if you are unable to use modular resources because you need to host `viam-server` on a non-Linux system or have an issue with compilation, you may need to [implement a custom component and register it on a server configured as a remote](/operate/reference/advanced-modules/custom-components-remotes/) on your machine.

## Package and deploy using Docker

In rare cases, you may need to package and deploy a module using Docker.
Use cases for this include:

> - Your module has complex system dependencies that cannot be installed on a machine.

Reviewer 2: *that cannot _easily_ be installed on a machine.

If they can be installed in docker, they can be installed on the machine. It just might be a giant pain.

Author: ```suggestion
- Your module has complex system dependencies that cannot be easily installed on a machine.
```

> - You have a large bundle and want to use layer caching to reduce the size of the download.

Reviewer 2: The world "bundle" seems odd, but I can't think of a better word. If no one else can think of a better word either, leave as-is.

Reviewer: I suppose that depends on what's meant by bundle? Does the layer caching help because an image with the same base container and possibly some layers can be used between different installed modules? That's the only way I can think this would be helpful.

Reviewer 2: The people who love docker will talk your ear off about how useful layer caching is: the idea is that if you have a bunch of small layers that add specific things to your docker image, you can reuse these layers in different projects to get different images without needing to create each thing multiple times. This is an argument for docker use in general, not something Viam-specific.

I think the bundle is supposed to refer to the collection of layers. I don't know enough about docker to know if that's the standard term; it might be. Even if it's not, I think the meaning is clear by the end of the sentence, I just stumbled a little in the middle.

Author: @abe-winter I borrowed your words that were perhaps not intended for official documentation. Is there a word other than "bundle" you'd recommend here?

Reviewer 3: maybe 'artifact'?

Reviewer: Right so it's useful if you have a large image and you use some of the same layers already on the machine so that they'll already be present. So that is the use case. Otherwise there wouldn't be anything reducing the size of the download. 

```suggestion
- You use a large container image and some layers are already used by your machine which means layer caching can reduce the size of the download.
```

- You have specific security requirements that are difficult to meet with the default module deployment.

> If you choose to deploy your module using Docker, we recommend creating a first run script or binary to run any necessary setup steps.

Reviewer 2: I'd like to better indicate that "first run" is a single thing, so no one mistakes it for, say, the first of many different run scripts or something. Would putting quotes around "first run" be appropriate? What about an underscore in first_run? I dunno, maybe leave as-is if that's standard in the docs.

Author: ```suggestion
If you choose to deploy your module using Docker, we recommend creating a "first run" script or binary to run any necessary setup steps.
```

{{% expand "Click for first run script instructions" %}}

1. Create a tarball that contains:

   - The module's entrypoint script or binary
   - A <file>meta.json</file>
   - A first run script or binary that will be executed during the setup phase, for example:

     ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
     #!/usr/bin/env bash

>      if [[ -n "$VIAM_TEST_FAIL_RUN_FIRST" ]]; then

Reviewer 2: I'm not familiar with this env var. Is it supposed to be in all first_run scripts? What does it do? Maybe I should add it to mine... :flushed:

Author: Great question--just pulled this in from the setup phase guide on confluence which in turn links to https://github.com/viam-labs/wifi-sensor/blob/7823b6ad3edcbbbf20b06c34b3181453f5f3f078/first_run.sh

Reviewer: I also am unclear on what this is for here. Can you add an explanation in? Does this get set to true if a previous run failed? Or is this jsut for testing?

Reviewer 2: In the RDK, this env var is mentioned in 2 places: a test and an example first_run.sh. I have cloned about 60 Viam-related repos locally, and it's not mentioned in any of the others. I've searched Github for it, and can only find copies of those 2 locations. I wonder if there was some plan to add this in, but it never happened and these are just the vestiges of a feature that never was.

I think we should remove this paragraph from the example.

Author: Thanks for looking around. Will remove.

>          echo "Sorry, I've failed you."

Reviewer: Does this error indicate an issue with the module? Maybe better to change to something that indicates where the error is

Author: I am not sure where it is. Since this is an edge case documented within a hidden page, I figured it made sense to act quickly and use the existing example provided by Maxim. Sounds like I'll remove the whole paragraph anyway.

         exit 1
     fi

>      docker pull mongo:6

Reviewer: So this pulls the image but then how does the container get run? I think we probably want to link to an example of that. Also 

And should this run with https://github.com/viam-soleng/viam-docker-manager or is this entirely separate?

Reviewer 2: I think this example is right: in first_run.sh, you pull the image, and then in your module's main executable, you run the image, and that second part should be obvious for anyone developing with docker. We could add in such an example, but it's trivial:
```bash
#!/bin/bash
set -euxo pipefail
# This example module runs inside the docker image you should have downloaded in first_run.sh
SOCKET_DIR=`dirname $1`
VIAM_DIR=`realpath ~/.viam`
exec docker run \
     --rm \
     --runtime=nvidia --gpus=all \
     -v /etc/passwd:/etc/passwd \
     -v /etc/group:/etc/group \
     -v $SOCKET_DIR:$SOCKET_DIR \
     -v $VIAM_DIR:$VIAM_DIR \
     ghcr.io/viam-modules/viam-mlmodelservice-triton/jetpack6:0.9.0 \
     LD_PRELOAD=libjemalloc.so.2 VIAM_MODULE_DATA="$VIAM_MODULE_DATA" /opt/viam/bin/viam_mlmodelservice_triton "$@" 2>&1
```

Reviewer 2: (that example is from the [Triton module](https://github.com/viam-modules/viam-mlmodelservice-triton/tree/main), and although there are lots of options specified for the `docker` line, anyone developing with docker should know which ones they need to set.)

Author: Okay, added above. Tested using the hello-world image, though it of course doesn't behave exactly like a real module that would continue to run. Did confirm that it ran the first_run and made a marker file.

     cat << EOF
     -------------------------------------
     Congratulations!
     The setup script ran successfully!
     This message is obnoxiously large for
     testing purposes.
     Sincerely,

>      First Run Script

Reviewer: Is it necessary to show a large message for the docs?

Author: I will reduce it. This was Max's original message and thought it might be okay to bring some amusement to the contents of an expander, further hidden within a hidden page, but no it doesn't need to be here.
```suggestion
     The setup script ran successfully!
```

     -------------------------------------
     EOF

     exit 0
     ```

   [This example Makefile on GitHub](https://github.com/viam-labs/wifi-sensor/blob/7823b6ad3edcbbbf20b06c34b3181453f5f3f078/Makefile) builds a module binary and bundles it along with a <file>meta.json</file> and first run script.<br><br>

1. Edit the <file>meta.json</file> to include a `first_run` field that points to the first run script or binary.

   ```json
   {
     ...
     "first_run": "first_run.sh"
   }
   ```

1. Configure your module on your machine in the same way as you would for a regular module.
   The first run script will execute once when `viam-server` receives a new configuration.
   It will only execute once per module or per version of the module.

1. (Optional) After a first run script runs successfully, Viam adds a marker file with a `.first_run_succeeded` suffix in the module’s data directory on disk.

>    It has the location and form: `/root/.viam/packages/data/module/<MODULE_ID>-<VERSION>-<ARCH>.first_run_succeeded`.

Reviewer 2: I suspect there should be a slash before `.first_run_succeeded`, but haven't checked on that personally. Maybe it's correct as-is.

Author: Hmm also from Maxim's guide--not sure if there's a reasonably practical way to check for sure

Author: Eep! I don't have any Docker modules to test this on... @cheukt since it looks like you've been maintaining the confluence instructions, can you confirm?

Reviewer 4: it is `unpackedModDir + FirstRunSuccessSuffix`, an example would be `.viam/packages/data/module/e76d1b3b-0468-4efd-bb7f-fb1d2b352fcb-viamrtsp-0_1_0-linux-amd64/bin.first_run_succeeded`
https://github.com/viamrobotics/rdk/blob/1af4d812c36e43c2f08b847290e1642489eb1df8/config/module.go#L279
